### PR TITLE
arch: arm: M0(+), SW_VECTOR_RELAY and VTOR

### PR DESCRIPTION
**Pull request by carlescufi**
VTOR and SW_VECTOR_RELAY interaction

### CODE REVIEW

**Comment by carlescufi** [comment]
CC @miyatsu @tagunil

File: arch/arm/core/cortex_m/Kconfig [hunk 1]
```diff
@@ -203,7 +203,8 @@ config SW_VECTOR_RELAY
 	prompt "Enable Cortex-M0 Vector Table soft relay"
 	default n
 	default y if BOOTLOADER_MCUBOOT
-	depends on CPU_CORTEX_M0 && !CPU_CORTEX_M0_HAS_VECTOR_TABLE_REMAP
+	depends on CPU_CORTEX_M0 && !(CPU_CORTEX_M_HAS_VTOR ||
```
Comment (ioannisg) [inline arch/arm/core/cortex_m/Kconfig]: @carlescufi please clarify: can the M0 have a VTOR register, only is this part of M0+?

File: arch/arm/core/cortex_m/Kconfig [hunk 2]
```diff
@@ -203,7 +203,7 @@ config SW_VECTOR_RELAY
 	prompt "Enable Cortex-M0 Vector Table soft relay"
 	default n
 	default y if BOOTLOADER_MCUBOOT
-	depends on CPU_CORTEX_M0 && !CPU_CORTEX_M0_HAS_VECTOR_TABLE_REMAP
+	depends on CPU_CORTEX_M0 && !(CPU_CORTEX_M_HAS_VTOR || CPU_CORTEX_M0_HAS_VECTOR_TABLE_REMAP)
```
Comment (ioannisg) [inline arch/arm/core/cortex_m/Kconfig]: @carlescufi please clarify: can the M0 have a VTOR register? Or is it part of M0+, and ARMv7?

Comment (carlescufi) [reply arch/arm/core/cortex_m/Kconfig]: The M0+ has a VTOR register, the M0 does not. Since we do not have an M0_PLUS Kconfig we use this.

Comment (carlescufi) [reply arch/arm/core/cortex_m/Kconfig]: ARMv7-M always has VTOR

Comment (nvlsianpu) [reply arch/arm/core/cortex_m/Kconfig]: LGTM

Comment (ioannisg) [reply arch/arm/core/cortex_m/Kconfig]: The M0 cannot have VTOR, the M0+ _may_ have VTOR.

**Comment by carlescufi** [comment]
@ioannisg thanks for the feedback, PR updated.

File: arch/arm/core/cortex_m/Kconfig [hunk 3]
```diff
@@ -203,7 +205,7 @@ config SW_VECTOR_RELAY
 	prompt "Enable Cortex-M0 Vector Table soft relay"
```
Comment (ioannisg) [inline arch/arm/core/cortex_m/Kconfig]: M0/M0+

File: arch/arm/core/cortex_m/Kconfig [hunk 4]
```diff
@@ -203,7 +205,7 @@ config SW_VECTOR_RELAY
...
 	default n
 	default y if BOOTLOADER_MCUBOOT
-	depends on CPU_CORTEX_M0 && !CPU_CORTEX_M0_HAS_VECTOR_TABLE_REMAP
+	depends on (CPU_CORTEX_M0 && !CPU_CORTEX_M0_HAS_VECTOR_TABLE_REMAP) || (CPU_CORTEX_M0PLUS && !CPU_CORTEX_M_HAS_VTOR)
```
Comment (ioannisg) [inline arch/arm/core/cortex_m/Kconfig]: Why CPU_CORTEX_M0_HAS_VECTOR_TABLE_REMAP is missing from the conditional of M0+?

Is it possible to have an M0+, no VTOR implemented, and no CPU_CORTEX_M0_HAS_VECTOR_TABLE_REMAP supported?

Comment (carlescufi) [reply arch/arm/core/cortex_m/Kconfig]: agreed, addressed

File: arch/arm/core/cortex_m/Kconfig (hunk 3 above)
Comment (carlescufi) [reply arch/arm/core/cortex_m/Kconfig]: agreed, addressed

**Comment by carlescufi** [comment]
@nvlsianpu @ioannisg I switched to use `ARMV6_M` instead, makes it simpler

**Review by nvlsianpu — APPROVED** [review]

**Review by ioannisg — APPROVED** [review]

**Comment by miyatsu** [comment]
Looks great. Only enables for SoC without HW relocation mechanisms.

**Review by galak — APPROVED** [review]

**Review by galak — APPROVED** [review]

**Pull request merged**